### PR TITLE
Assignment client keep a spare

### PR DESCRIPTION
**Pull request by sethalves**
AssignmentClientMonitor talks with its children over udp.  Every 6 seconds, if none are idle it spawns a new AssignmentClient.  If 2 or more are idle, it kills one.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 270d509a1284c3b8fb3ca07e6ea077e9c7f49eb1 available at: https://gist.github.com/fe4dbcff7ce59d82afef<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bbb6e048f5d77eeafdd9e5e78ccf8c653162b29c available at: https://gist.github.com/175a5fb46fd60c340603<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.dmg

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4324.mov

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4324.mov

File: assignment-client/src/AssignmentClientApp.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        monitor.exec();
+    } else {
+        AssignmentClient client(argc, argv);
+        client.exec();
+    }
+}
+
+
+AssignmentClientApp::~AssignmentClientApp() {
+}
```
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.cpp]: remove this now that it is empty

File: assignment-client/src/AssignmentClientApp.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#include <QApplication>
+
+class AssignmentClientApp : public QCoreApplication {
+    Q_OBJECT
+public:
+    AssignmentClientApp(int argc, char* argv[]);
+    ~AssignmentClientApp();
```
Comment (birarda) [inline assignment-client/src/AssignmentClientApp.h]: no longer needed

File: assignment-client/src/AssignmentClientMonitor.cpp [hunk 3]
```diff
@@ -53,46 +70,129 @@ AssignmentClientMonitor::~AssignmentClientMonitor() {
...
-        it = _childProcesses.erase(it);
-    }
+    auto nodeList = DependencyManager::get<NodeList>();
+
+    nodeList->eachNode([&](const SharedNodePointer& node){
+            qDebug() << "asking child" << node->getUUID() << "to exit.";
+            node->activateLocalSocket();
+            QByteArray diePacket = byteArrayWithPopulatedHeader(PacketTypeStopNode);
+            nodeList->writeUnverifiedDatagram(diePacket, *node->getActiveSocket());
+        });
```
Comment (birarda) [inline assignment-client/src/AssignmentClientMonitor.cpp]: indent for this block is off

File: assignment-client/src/AssignmentClientMonitor.cpp [hunk 4]
```diff
@@ -53,46 +70,129 @@ AssignmentClientMonitor::~AssignmentClientMonitor() {
...
+
+    nodeList->removeSilentNodes();
+
+    nodeList->eachNode([&](const SharedNodePointer& node){
+            AssignmentClientChildData *childData = static_cast<AssignmentClientChildData*>(node->getLinkedData());
+            if (childData->getChildType() == "none") {
+                spareCount ++;
+                aSpareId = node->getUUID();
+            }
+        });
```
Comment (birarda) [inline assignment-client/src/AssignmentClientMonitor.cpp]: indent for this block is off

File: assignment-client/src/AssignmentClientMonitor.cpp [hunk 5]
```diff
@@ -53,46 +70,129 @@ AssignmentClientMonitor::~AssignmentClientMonitor() {
...
+    }
+}
+
+
+AssignmentClientChildData::AssignmentClientChildData(QString childType) {
+    _childType = childType;
+}
+
+
+AssignmentClientChildData::~AssignmentClientChildData() {
```
Comment (birarda) [inline assignment-client/src/AssignmentClientMonitor.cpp]: remove this if the implementation is empty

File: assignment-client/src/AssignmentClientMonitor.cpp [hunk 6]
```diff
@@ -53,46 +70,129 @@ AssignmentClientMonitor::~AssignmentClientMonitor() {
...
+                // have the NodeList attempt to handle it
+                nodeList->processNodeData(senderSockAddr, receivedPacket);
+            }
+        }
+    }
+}
+
+
+AssignmentClientChildData::AssignmentClientChildData(QString childType) {
+    _childType = childType;
```
Comment (birarda) [inline assignment-client/src/AssignmentClientMonitor.cpp]: use the constructor initializers for the `_childType` member variable

_childType(childType) before the empty method definition

File: assignment-client/src/AssignmentClientMonitor.h [hunk 7]
```diff
@@ -15,25 +15,43 @@
...
+
+    QString getChildType() { return _childType; }
+    void setChildType(QString childType) { _childType = childType; }
+
+    // implement parseData to return 0 so we can be a subclass of NodeData
+    int parseData(const QByteArray& packet) { return 0; }
+
+ private:
+    QString _childType;
+};
```
Comment (birarda) [inline assignment-client/src/AssignmentClientMonitor.h]: is there a reason this class shouldn't be in it's own header and implementation file

File: libraries/networking/src/LimitedNodeList.cpp [hunk 8]
```diff
@@ -669,3 +669,41 @@ void LimitedNodeList::sendHeartbeatToIceServer(const HifiSockAddr& iceServerSock
...
+        sharedPortMem->unlock();
+        
+        qDebug() << "Wrote local listening port" << localPort << "to shared memory at key" << key;
+    } else {
+        qWarning() << "Failed to create and attach to shared memory to share local port with assignment-client children.";
+    }
+}
+
+
+bool LimitedNodeList::getLocalPortFromSharedMemory(const QString key, QSharedMemory* &sharedMem,
```
Comment (birarda) [inline libraries/networking/src/LimitedNodeList.cpp]: I think this method needs a different name - it currently suggests that you're getting the local listening port for this NodeList from shared memory

Comment (birarda) [inline libraries/networking/src/LimitedNodeList.cpp]: the `QSharedMemory` pointer does not need to be passed by reference here

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed without errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4324.mov

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0b9b7f06cc8088b530b3b99d9d23d0f74322ed93 available at: https://gist.github.com/4b908568da6487ded946<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.dmg

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4324.mov

**Comment by hifi-gustavo** [comment]
OS X Yosemite functional test completed with errors
Download recording here:
https://s3.amazonaws.com/hifi-public/pr-builds/recordings/4324.mov

**Comment by hifi-gustavo** [comment]
Build console output for commit: f5a7a19bc3f239e04836f98d7c8fc0d24cfe78de available at: https://gist.github.com/5b41a5cf96ac7d4b3255<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 88920ad3aadd67e7c50735138c216f8d0b6503ad available at: https://gist.github.com/56eefb515ce8e21337d6<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4324.dmg

**Comment by OmegaHeron** [comment]
And due to the spare AC process the log is flooded with;

`[DEBUG] [02/24 14:54:59] Received a request for assignment type 7 from 127.0.0.1:49959
[DEBUG] [02/24 14:54:59] Unable to fulfill assignment request of type 7 from 127.0.0.1:49959`

At 1 second intervals .

**Comment by birarda** [comment]
@sethalves how do you want to handle that? We could make that message the "supress" type and combine it into one message a minute or something?

**Pull request merged**